### PR TITLE
Wildan/fix/21518/client pusher method

### PR DESCRIPTION
**Pull request by wildan-m**
@johncschuster , @allroundexperts
### Details

Fix leaving a room from multiple devices and fix regression when splitting the bill

### Fixed Issues

$ https://github.com/Expensify/App/issues/21518
$ https://github.com/Expensify/App/issues/22964
PROPOSAL: https://github.com/Expensify/App/issues/21518#issuecomment-1621007692

### Leave Room Test
1. Open a room on the web and on another device (e.g. Android) using the same account
2. Click room details and click `Leave Room` on one of the devices (e.g. web)
3. The user should leave the room on both devices

### Split bill test
1. Go to Website
2. Split bill.
3. Verify that some of the recipients are a recipient with no interaction
4. Click on preview.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Open a room on the web and on another device (e.g. Android) using the same account
2. disconnect the internet on both
3. Click room details and click `Leave Room` on one of the devices (e.g. web)
5. At this point the user should only leave a room in the web version only.
6. re-connect the internet connection
7. The user should leave the room on both devices

### QA Steps

1. Open a room on the web and on another device (e.g. Android) using the same account
2. Click room details and click `Leave Room` on one of the devices (e.g. web)
3. The user should leave the room on both devices

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/11847279/8489739f-8cb1-4a51-9b14-f798079a1fa9

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/11847279/2094d134-5023-4c91-8bd5-8cb5817cd4ac

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/11847279/f3ca72fc-3c14-4a3d-9cc4-20ce196e03c0

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/11847279/07aeac99-a36c-4364-a919-5e0cb723ce62

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/11847279/f575183a-880a-46ba-8c8b-f3ad5c8fd003

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/11847279/863ff926-d808-4425-bd5d-d4357fc9b6fc

</details>

### CODE REVIEW

**Comment by wildan-m** [comment]
@allroundexperts sure. Done.

**Comment by wildan-m** [comment]
@allroundexperts any feedback on the implementation?

**Comment by wildan-m** [comment]
@allroundexperts sure. done.

**Comment by wildan-m** [comment]
@allroundexperts not sure what causes some checks to fail. seems not related to this PR

**Comment by wildan-m** [comment]
resolved after [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1692680548820779).

**Comment by deetergp** [comment]
@wildan-m It looks like `ONYXKEYS.js` is modified in your branch but has been replaced by `ONYXKEYS.ts` in `main`. Can you please resolve the conflict and push it? 

@allroundexperts let's try to get this reviewed before the merge conflict merry-go-round comes back around to haunt Wildan again? 😅

**Comment by wildan-m** [comment]
@deetergp, @allroundexperts done.

**Comment by wildan-m** [comment]
@allroundexperts,  @deetergp just resolved conflict again. any updates?

**Comment by luacmartins** [comment]
@allroundexperts bump for review

**Comment by wildan-m** [comment]
@allroundexperts bump

File: src/libs/actions/Report.js
Comment (luacmartins) [inline src/libs/actions/Report.js:195]:
```suggestion
 * Initialize our pusher subscriptions to listen for someone leaving a room.
```

File: src/libs/actions/Report.js
Comment (luacmartins) [inline src/libs/actions/Report.js:137]: Can we just rename the existing method to something like `getNormalizedStatus`? There's no need to duplicate that logic

File: src/libs/actions/Report.js
Comment (luacmartins) [inline src/libs/actions/Report.js:940]:
```suggestion
 * Broadcasts to the report's private pusher channel whether a user is leaving a report
```

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [inline src/pages/home/ReportScreen.js:91]:
```suggestion
    /** Whether user is leaving the current report */
```

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [inline src/pages/home/ReportScreen.js:293]:
```suggestion
        // Navigate to the Concierge chat if the room was removed from another device (e.g. user leaving a room)
```

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [inline src/pages/home/ReportScreen.js:295]:
```suggestion
```

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [inline src/pages/home/ReportScreen.js:297]:
```suggestion
```

**Review by luacmartins — COMMENTED** [review]
Looking good. Left a couple of comments

File: src/libs/actions/Report.js
Comment (wildan-m) [reply src/libs/actions/Report.js:195]: forget to change that 😂

File: src/libs/actions/Report.js
Comment (wildan-m) [reply src/libs/actions/Report.js:137]: @luacmartins did you mean to make typing and leaving the same method?

```
function getNormalizedStatus(typingStatus) {
    let normalizedStatus = typingStatus;

    if (_.first(_.keys(typingStatus)) === 'userLogin') {
        normalizedStatus = {[typingStatus.userLogin]: true};
    }

    return normalizedStatus;
}
```

File: src/pages/home/ReportScreen.js
Comment (wildan-m) [reply src/pages/home/ReportScreen.js:295]: I think this tiny comment is required to explain why we have two separate logic

File: src/libs/actions/Report.js
Comment (luacmartins) [reply src/libs/actions/Report.js]: yup, you got it right!

File: src/pages/home/ReportScreen.js
Comment (luacmartins) [reply src/pages/home/ReportScreen.js:295]: I think it's fine without, but I won't block on it

**Review by luacmartins — DISMISSED** [review]

**Comment by wildan-m** [comment]
@allroundexperts let me check it. Before each test can you make sure all of the opened instances refreshed? 

How often is it?

**Comment by wildan-m** [comment]
@allroundexperts I've tried 5 rooms with various types, and no navigation to the concierge issue. But, I noticed that the concierge screen showed twice in iOS native. that might caused by a different issue, I'm investigating it. Meanwhile, can you try again with the newly merged commit?

https://github.com/Expensify/App/assets/11847279/4dd8ac62-8eb8-4c9b-a506-9bc892fd0221

**Comment by wildan-m** [comment]
@allroundexperts, I've updated the code to send the leaving status broadcast before calling the LeavingRoom API. There is a race condition between the onyx report becoming null and the leavingStatus being updated.

File: src/libs/actions/Report.js
Comment (wildan-m) [reply src/libs/actions/Report.js]: this one fixed, not sure why your last week comment just showed up today to me

File: src/libs/actions/Report.js
Comment (wildan-m) [reply src/libs/actions/Report.js]: @allroundexperts 
did you mean to change
```
_.first(_.keys(leavingStatus)) === 'userLogin'
```
to
```
const keys = Object.keys(status);
const isFirstKeyUserLogin = keys.length > 0 && keys[0] === 'userLogin';
```

?

File: src/libs/actions/Report.js
Comment (wildan-m) [reply src/libs/actions/Report.js:192]: I cloned it from typing, alternatively, we can the below code if required, but a little bit longer
```
const normalizedLeavingStatus = getNormalizedLeavingStatus(leavingStatus);
const accountIDOrLogin = normalizedLeavingStatus ? Object.keys(normalizedLeavingStatus)[0] : null;
```

**Comment by wildan-m** [comment]
@allroundexperts bump!

**Comment by luacmartins** [comment]
@wildan-m we have conflicts

**Comment by wildan-m** [comment]
@luacmartins the conflict has been resolved

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
@allroundexperts is this still on your list for today?

**Review by allroundexperts — APPROVED** [review]
Testing well now!

**Review by luacmartins — APPROVED** [review]

File: src/pages/home/ReportScreen.js
Comment (wildan-m) [reply src/pages/home/ReportScreen.js:340]: @fedirjh Thanks for your feedback. We did it [here](https://github.com/Expensify/App/blob/6cc905388954bdcec3be705b743595f0a3f41a03/src/libs/actions/Report.js#L192-L194) and [here](https://github.com/Expensify/App/blob/6cc905388954bdcec3be705b743595f0a3f41a03/src/libs/actions/Report.js#L242-L244).

The actual issue is this function:
https://github.com/Expensify/App/blob/6cc905388954bdcec3be705b743595f0a3f41a03/src/pages/home/ReportScreen.js#L128-L131

it returned 'null' string instead of an actual `null`.

I'd recommend to change it to:

```
function getReportID(route) {
    const reportID = lodashGet(route, 'params.reportID', null);
    return reportID !== null ? String(reportID) : null;
}
```

@luacmartins, @allroundexperts, considering that this has passed the regression period, the buggy function is not directly related to this PR, and any change would require re-testing. What should we do in this situation? Will there be any adjustments?

Comment (wildan-m) [reply src/pages/home/ReportScreen.js:340]: @fedirjh seems you've resolved it [here](https://github.com/Expensify/App/pull/29590). thank you

Comment (Pujan92) [reply src/pages/home/ReportScreen.js:340]: I think we also need to check onyx report `reportId` availability to avoid https://github.com/Expensify/App/issues/38174

**Pull request merged**
